File: pi.py
#!/usr/bin/env python3

#####################################################
#                                                   #
# License: Apache License 2.0                       #
# Author: Dario Ostuni <another.code.996@gmail.com> #
#                                                   #
#####################################################

import sys

def fatt(n):
    tmp = 1
    for i in range(1, n + 1):
        tmp *= i
    return tmp

def term(n):
    num = ((-1)**n)*fatt(4*n)*(21460*n+1123)
    den = (fatt(n)**4)*(14112**(2*n))
    return num, den
def mcd(a, b):
    if a < b:
        a, b = b, a
    while b != 0:
        a, b = b, a % b
    return a

if len(sys.argv) == 2:
    cfr = int(sys.argv[1])
else:
    cfr = int(input("Number of digits: "))
prec = cfr // 5 + 10
num = 0
den = 1
for i in range(prec):
    tmp_n, tmp_d = term(i)
    num = num * tmp_d + den * tmp_n
    den = den * tmp_d
    gdc = mcd(num, den)
    num //= gdc
    den //= gdc

num, den = den * 3528, num

num -= 3 * den
print("3.", end='')
for i in range(cfr):
    num *= 10
    print(num // den, end='')
    num %= den
print(flush=True)

File: pi.rs
// Licence: Apache 2.0
// Copyright (2017): Dario Ostuni <dario.ostuni@gmail.com>

fn main()
{
    let pi = 3528.0 /
             (0..)
        .take(3)
        .map(|n: i64| {
            ((1..)
                .take(n as usize)
                .fold(1, |acc, _| -1 * acc) *
             (1..)
                .take(4 * n as usize)
                .product::<i64>() * (21460 * n + 1123)) as f64 /
            ((1..)
                .take(4)
                .fold(1,
                      |acc, _| {
                          (1..)
                              .take(n as usize)
                              .product::<i64>() * acc
                      }) *
             (1..)
                .take(4 * n as usize)
                .fold(1, |acc, _| 4 * acc) *
             (1..)
                .take(2 * n as usize)
                .fold(1, |acc, _| 882 * acc)) as f64
        })
        .sum::<f64>();
    println!("{}", pi);
}
